### DISCUSSION TITLE
Help with undistortion of b0

### DISCUSSION THREAD

**Discussion by leandrolecca** [dipy/dipy#2811] [3. Community Lounge]
Dear community,

I'm using the [predict() function of Synb0](https://github.com/dipy/dipy/blob/aa39ed3fb0a7397215b6e9b9ef746496a9e4091c/dipy/nn/synb0.py#L418) to get b0 undistorted images. I have a couple of questions:

- Can I input a non-skull stripped T1? Does the function perform skull stripping for me?
- Why does the function require a 77x91x77 voxel input? If not, it fails. Could I edit the shape of the [UNet3D](https://github.com/dipy/dipy/blob/aa39ed3fb0a7397215b6e9b9ef746496a9e4091c/dipy/nn/synb0.py#L368)?

I'm adding a very preliminar output (T1 wasn't skull-stripped). I'd like to get a larger voxel size output. 

Thanks in advance!

![image](https://github.com/dipy/dipy/assets/49255427/a1320a54-2e5c-4adb-b2c8-4917f6371d74)

**Comment by skoudoro** [comment]
Hi @leandrolecca,

> Can I input a non-skull stripped T1? Does the function perform skull stripping for me?

No, You need to do it yourself before. You can use evac+ algorithm or median_otsu in DIPY

> Why does the function require a 77x91x77 voxel input? If not, it fails. Could I edit the shape of the [UNet3D](https://github.com/dipy/dipy/blob/aa39ed3fb0a7397215b6e9b9ef746496a9e4091c/dipy/nn/synb0.py#L368)?

Synb0 is a whole pipeline why has been trained on a specific image size. In DIPY, we just have a part of the pipeline. The rest will come as soon as possible. 

For sure @pjsjongsung and @frheault can tell you more about the original implementation and the current one in DIPY. 

Thank you for your questions and feedback and they will answer you as soon as they can.

**Comment by pjsjongsung** [comment]
Hi @leandrolecca! Thank you for your interest. If you are interested in the whole pipeline, Synb0 has its own docker image https://github.com/MASILab/Synb0-DISCO

**Comment by smeisler** [comment]
Hi all, I would like to continue this thread.

As it stands it appears that DIPY implements the last stage of the pipeline and presumes inputs follow the the pipeline in the[ Synb0 paper](https://www.sciencedirect.com/science/article/pii/S0730725X18306179?via%3Dihub). However, after looking at the DIPY documentation, it is unclear what this means precisely. I would like to add some synb0 authors here so we can resolve some of these issues - I am having trouble tagging them here but will alert them to this thread. 
___
From the paper:

> To apply the trained networks, an MPRAGE dataset without a paired multi-shot EPI scan is affine registered (12 degree of freedom using FSL flirt [[24](https://www.sciencedirect.com/science/article/pii/S0730725X18306179?via%3Dihub#bb0120)]) to the MNI-152 1 mm isotropic T1-weighted atlas. All overlapping sets of three contiguous slices for the scan are written out (193, 229, and 193 stacks of 2D images per subject, axial, coronal, sagittal, respectively).

For anyone who needed clarification, this is **`MNI152NLin6Sym`** and not **`MNI152NLin2009cAsym`**.

For DIPY developers @pjsjongsung : In DIPY why does the image have to be 77X91X77? Unless I am misunderstanding something that appears to be in contradiction with what is described above, unless only a particular part of the neural net is implemented in DIPY. If so, what part is it?
___
Similarly, I would like to confirm that the following needs to take place before entering the neural net, and that this is indeed what is expected before DIPY.

1) Skull-stripped
2) Normalized
3) Aligned to one another in MNI152NLin6Sym space (1mm voxel for T1, **what about b0**)?
___
Next, this argument is used in Synb0-DISCO and I cannot find an analogous step in DIPY:
> acqparams.txt: A text file that describes the acqusition parameters, and is described in detail on the FslWiki for topup (https://fsl.fmrib.ox.ac.uk/fsl/fslwiki/topup). Briefly, it describes the direction of distortion and tells TOPUP that the synthesized image has an effective echo spacing of 0 (infinite bandwidth). An example acqparams.txt is displayed below, in which distortion is in the second dimension, note that the second row corresponds to the synthesized, undistorted, b0:
> ```
> $ cat acqparams.txt 
> 0 1 0 0.062
> 0 1 0 0.000
> 
> ```
Is this information not used in DIPY, or is it presumed from something?

Thanks for your help,
Steven

**Comment by schillkg** [comment]
Hi all - I was just directed to this and hope to answer all the questions.

Network: First, the newest synthesis is based on a 3D Unit described here: https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0236418

I think this may clarify some things:
The end-to-end pipeline includes (1) preparing the input distorted data (more on this later, but basically aligns T1 and b0 to MNI space) (2) running inference to synthesize an undistorted b0 (3) averaging the 5 synthesized images we get (5 networks or 5 folds) (4) applying the inverse transform to get back to native space (5) slight smoothing of the synthesized image (this regularized topup a bit) (6) concatenate the distorted and undistorted images to act as input to topup using the acqparams file described above and (7) run topup to estimate the field!

This process is contained in our pipeline.sh file (https://github.com/MASILab/Synb0-DISCO/blob/master/src/pipeline.sh)

Now the #1 prepare input step includes: (1) image normalization from free surfer (normalizes so WM has a value ~110 or so) (2) skull strip brain (we only do this to prepare for epi_reg which requires a skull stripped T1) (3) register b0 to T1 (using epi_reg) (4) convert FSL transform to ANTS transform (so we can concatenate transforms in a later step) (5) ANTS register T1 to atlas (6) apply transforms to get both b0 and T1 into MNI space. Now we go ahead and apply step #2 from above (image synthesis). 

This input preparation is here: (https://github.com/MASILab/Synb0-DISCO/blob/master/data_processing/prepare_input.sh)

We happened to chose the mni_icbm152_t1_tal_nlin_asym_09c.nii.gz atlas at 2.5mm resolution, so that is why the input to the synthesis requires a hard-coded (and weird) dimensions. However, the raw input (t1 and b0) can be any resolution and/or field of view. 

So yes - the full end-to-end pipeline can include either skull or skull stripped, and can be any resolution. Happy to try to answer any more questions - or questions specific to the Dipy implementation.

**Comment by smeisler** [comment]
Great, thanks @schillkg!!

So now what I’ve gathered is that DIPY is just doing the final step of just synthesis. It seems like, if one only wanted to use DIPY to implement this, one would need to have a t1 that has been  (1) n4 normalized via freesurfer  (2) skull stripped (3) in MNI152nLin2009cAsym 2.5mm isotropic space. And a b0 that has been (1) aligned to that t1 in same space and same voxel resolution. The commands in the provided script (or similar ones) can be used to achieve this.

**Comment by Garyfallidis** [comment]
@smeisler you may want to chat with @sreekarchigurupati who is working with Kurt and others to simplify the process.

**Comment by Garyfallidis** [comment]
We are looking for more people to test the upcoming versions. Maybe a great opportunity for a collaboration.

**Reply by smeisler** [reply]
Hi @Garyfallidis, good to know. I would definitely be interested in helping to test this (and build too, if needed)!

**Comment by frheault** [comment]
Hi @schillkg @Garyfallidis @smeisler !

This is a working version of a wrapper script for this code, it is still a PR in our library, but this is pretty much exactly what you want/need!
https://github.com/scilus/scilpy/pull/734 

(At least related to using the model)
